soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java

```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
 			List<SootMethod> methodCopyList = new ArrayList<>(sootClass.getMethods());
 			for (SootMethod sootMethod : methodCopyList) {
 				if (sootMethod.isConcrete()) {
-					final Body body = sootMethod.retrieveActiveBody();
-					final LocalGenerator lg = Scene.v().createLocalGenerator(body);
+					final Body body = sootMethod.retrieveActiveBody();				
 
 					// Mark the method as processed
 					if (!processedMethods.add(sootMethod))
 						continue;
 
 					for (Iterator<Unit> unitIter = body.getUnits().snapshotIterator(); unitIter.hasNext();) {
-						Stmt stmt = (Stmt) unitIter.next();
-
-						if (stmt.containsInvokeExpr()) {
-							SootMethod callee = stmt.getInvokeExpr().getMethod();
-
-							// For Messenger.send(), we directly call the respective handler
-							if (callee == smMessengerSend) {
-								Set<SootClass> handlers = MessageHandler.v().getAllHandlers();
-								for (SootClass handler : handlers) {
-									Local handlerLocal = lg.generateLocal(handler.getType());
-
-									Unit newU = Jimple.v().newAssignStmt(handlerLocal,
-											Jimple.v().newNewExpr(handler.getType()));
-									newU.addTag(SimulatedCodeElementTag.TAG);
-									body.getUnits().insertAfter(newU, stmt);
-									instrumentedUnits.put(body, newU);
-
-									SootMethod initMethod = handler.getMethod("void <init>()");
-									Unit initU = Jimple.v().newInvokeStmt(
-											Jimple.v().newSpecialInvokeExpr(handlerLocal, initMethod.makeRef()));
-									initU.addTag(SimulatedCodeElementTag.TAG);
-									body.getUnits().insertAfter(initU, newU);
-									instrumentedUnits.put(body, initU);
+						Unit unit = unitIter.next();
+						Stmt stmt = (Stmt) unit;
+						//+++ collect handler fields signature	
+						if (stmt.containsFieldRef() && stmt.getFieldRef().getType().toString().contains("android.os.Handler")) {
+							if(stmt.getUseBoxes().size() > 1 && handlerInner.get(stmt.getFieldRef().getField().getSignature()) == null)
+								handlerInner.put(stmt.getFieldRef().getField().getSignature(),stmt.getUseBoxes().get(1).getValue().getType().toString());
+							// use value as locator
+							handlerClass.put(sootClass.getName() + ":" + stmt.getDefBoxes().get(0).getValue(),stmt.getFieldRef().getField().getSignature());
+						}
+							
+						if(stmt.containsInvokeExpr() && stmt.getInvokeExpr().getMethodRef().getSignature().contains("android.os.Handler")) {
+							//Soot has a access$XXX method that returns private fields in outer class
```

Comment:
The ``access`` methods are part of the bytecode. They are generated by the Java compiler to allow access to private methods from inner classes (and some other special cases). They are not an artifact of Soot. The code must work regardless of whether there is an ``access`` method or a direct call. Do not rely on the structure of the ``access`` method.

Reply:
Thanks for your patient reply, sorry for not getting back to you sooner. 
This code is meant to deal with private handlers. Removing it would be fine. 
BTW, if not rely on the structure of the `access` method, how does soot deal with Private handlers/inner classes?